File: client/lib/models/lottery.dart
import 'user.dart';

class LotteryEvent {
  final int id;
  final String title;
  final String description;
  final String prizeName;
  final DateTime drawTime;
  final int status; // 0: Upcoming/Ongoing, 1: Drawn
  final int? winnerId;
  final User? winner;

  LotteryEvent({
    required this.id,
    required this.title,
    required this.description,
    required this.prizeName,
    required this.drawTime,
    required this.status,
    this.winnerId,
    this.winner,
  });

  factory LotteryEvent.fromJson(Map<String, dynamic> json) {
    return LotteryEvent(
      id: json['id'],
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      prizeName: json['prize_name'] ?? '',
      drawTime: DateTime.parse(json['draw_time']),
      status: json['status'] ?? 0,
      winnerId: json['winner_id'],
      winner: json['winner'] != null ? User.fromJson(json['winner']) : null,
    );
  }
}

File: server/internal/models/lottery.go
package models

import (
	"time"
)

type LotteryEvent struct {
	ID          uint                 `gorm:"primarykey" json:"id"`
	CreatedAt   time.Time            `json:"created_at"`
	UpdatedAt   time.Time            `json:"updated_at"`
	Title       string               `gorm:"size:100;not null" json:"title"`
	Description string               `gorm:"type:text" json:"description"`
	PrizeName   string               `gorm:"size:100;not null" json:"prize_name"`
	DrawTime    time.Time            `json:"draw_time"`
	Status      int                  `gorm:"default:0" json:"status"` // 0: Upcoming/Ongoing, 1: Drawn
	WinnerID    *uint                `json:"winner_id"`
	Winner      *User                `gorm:"foreignKey:WinnerID" json:"winner,omitempty"`
	Participants []LotteryParticipant `gorm:"foreignKey:LotteryID" json:"participants,omitempty"`
}

type LotteryParticipant struct {
	ID        uint      `gorm:"primarykey" json:"id"`
	CreatedAt time.Time `json:"created_at"`
	LotteryID uint      `gorm:"uniqueIndex:idx_lottery_user;not null" json:"lottery_id"`
	UserID    uint      `gorm:"uniqueIndex:idx_lottery_user;not null" json:"user_id"`
	Weight    int       `gorm:"default:1;not null" json:"weight"`
	User      User      `gorm:"foreignKey:UserID" json:"user"`
}
